### PR TITLE
Improve doc

### PR DESCRIPTION
**Pull request by markuspg**
This extends and updates the configuration. Additionally the copyright notices are completed and updated and some file mode and line ending issues are being cleaned up.

### CODE REVIEW

File: lib/rutema/core/engine.rb [hunk 1]
```diff
@@ -29,10 +45,16 @@ def initialize configuration
       @dispatcher=Dispatcher.new(@queue,configuration)
       @configuration=configuration
     end
-    #Parse, run, report
+
+    ##
+    # Parse the test specifications, execute the test(s) and report the results
+    #
+    # * +test_identifier+ - an optional identifier of a single test case to be
+    #   executed, this cannot be an arbitrary one but must still be contained
```
Comment (mrkita) [inline lib/rutema/core/engine.rb:53]: I'm not really following the intention here for what you want to say about test_identifier. In particular about what you mean about not being arbitrary.

File: lib/rutema/core/engine.rb [hunk 2]
```diff
@@ -41,15 +63,22 @@ def run test_identifier=nil
       else
         @runner.setup=setup
         @runner.teardown=teardown
-        #running - at this point we've done any and all checks and we're stepping on the gas
+        # running - at this point all checks are done and the tests get busy
```
Comment (mrkita) [inline lib/rutema/core/engine.rb:66]: tests get busy -> tests are active
?

File: lib/rutema/core/engine.rb [hunk 3]
```diff
@@ -41,15 +63,22 @@ def run test_identifier=nil
...
       @dispatcher.report(tests)
     end
-    #Parse a single test spec or all the specs listed in the configuration
+
+    ##
+    # Parse a single test specification or all the specs given by the
+    # configuration
+    #
+    # * +test_identifier+ - an optional identifier of a single test case to be
+    #   executed, this cannot be an arbitrary one but must still be contained
```
Comment (mrkita) [inline lib/rutema/core/engine.rb:80]: See above

File: lib/rutema/core/engine.rb [hunk 4]
```diff
@@ -128,17 +192,41 @@ def run_test specification
...
         return klass.new(configuration)
       end
       return nil
     end
+
+    ##
+    # Check if the given test identifier belongs to the normal test cases or to
+    # one of the special ones (the test (suite) setups and teardowns)
+    #
+    # * +test_identifier+ - the test identifier which shall be checked if it
```
Comment (mrkita) [inline lib/rutema/core/engine.rb:215]: the test identifier to check against membership in the normal or special test case identifier sets.

File: lib/rutema/core/engine.rb [hunk 5]
```diff
@@ -128,17 +192,41 @@ def run_test specification
...
       return @configuration.tests.include?(full_path) || is_special?(test_identifier) 
     end
+
+    ##
+    # Check if the given test identifier is a (suite) setup or teardown one
+    #
+    # This checks if the given identifier was configured as (suite) setup or
+    # teardown within the rutema run's configuration.
+    #
+    # * +test_identifier+ - the test identifier which shall be checked for being
```
Comment (mrkita) [inline lib/rutema/core/engine.rb:228]: see above

File: lib/rutema/core/engine.rb [hunk 6]
```diff
@@ -169,12 +270,25 @@ def initialize queue,configuration
...
       @configuration=configuration
     end
-    #Call this to establish a queue with the given identifier
+
+    ##
+    # Call this to establish a queue with the given identifier
+    #
+    # This method will create a new queue within the Dispatcher instance into
+    # which data from the incoming queue from the Engine instance will be
+    # demultiplexed too.
```
Comment (mrkita) [inline lib/rutema/core/engine.rb:279]: dispatched to

File: lib/rutema/core/engine.rb [hunk 7]
```diff
@@ -169,12 +270,25 @@ def initialize queue,configuration
...
+    #
+    # * +identifier+ - a unique identifier for the queue. If two identifiers
+    #   collide the later added subscriber will take the place of the earlier
```
Comment (mrkita) [inline lib/rutema/core/engine.rb:282]: the new subscriber will replace the earlier one

File: lib/rutema/core/engine.rb [hunk 8]
```diff
@@ -169,12 +270,25 @@ def initialize queue,configuration
...
+    #   one
     def subscribe identifier
       @queues[identifier]=Queue.new
       return @queues[identifier]
     end
-    
+
+    ##
+    # Start #update threads of all event/streaming reporters and then start own
```
Comment (mrkita) [inline lib/rutema/core/engine.rb:290]: and then start a new locally managed thread

File: lib/rutema/core/engine.rb [hunk 9]
```diff
@@ -200,7 +321,12 @@ def exit
         Thread.kill(@thread)
       end
     end
+
     private
+
+    ##
+    # As long as the incoming queue is not empty dispatch its messages to all
```
Comment (mrkita) [inline lib/rutema/core/engine.rb:328]: Dispatch messages from the incoming queue to all subscribers until the queue is empty.

File: lib/rutema/core/engine.rb [hunk 10]
```diff
@@ -210,18 +336,35 @@ def flush
         end
       end
     end
+
+    ##
+    # Instantiate a new reporter instance and pass the configuration and the
+    # Dispatcher instance this method is called upon to
```
Comment (mrkita) [inline lib/rutema/core/engine.rb:342]: this dispatcher instance.

File: lib/rutema/parsers/xml.rb [hunk 11]
```diff
@@ -116,17 +143,38 @@ def parse_step xmltxt
...
         return step
       end
+
+      ##
+      # Add an attribute of a given name with the given value to a specification
+      # element
+      #
+      # * +element+ - the specification element the attribute shall be added to
+      # * +attr+ - the name of the attribute which shall either be created or
+      #   whose current value shall be override
```
Comment (mrkita) [inline lib/rutema/parsers/xml.rb:153]: will be overriden

File: examples/README.md [hunk 12]
```diff
@@ -1,46 +1,46 @@
...
+````
+config/ 
+specs/
+scripts/
+suites/
+````
+
+Configuration files are stored in config/ and suites/ and have per convention the extension .rutema.
+
+Note that there is no difference between a configuration file with tool parameters and a configuration file that defines a test suite. The difference is by convention:
```
Comment (mrkita) [inline examples/README.md:25]: Maybe we can get a description of "tool parameters" before this

File: examples/README.md [hunk 13]
```diff
@@ -1,46 +1,46 @@
...
+
+*Tool, path and test environment specific configuration is saved in config/ and the collection of tests to run per configuration in suites/*
+
+## rutema tooling
+
+The configuration system allows us to replace every part of the rutema engine. They are only three major parts anyway: the parser, the runner and the reporter. The parser must be specified, it handles the specific testing language you will develop for the project.
```
Comment (mrkita) [inline examples/README.md:31]: *There are three major parts

File: examples/README.md [hunk 14]
```diff
@@ -1,46 +1,46 @@
...
+
+Note that there is no difference between a configuration file with tool parameters and a configuration file that defines a test suite. The difference is by convention:
+
+*Tool, path and test environment specific configuration is saved in config/ and the collection of tests to run per configuration in suites/*
+
+## rutema tooling
+
+The configuration system allows us to replace every part of the rutema engine. They are only three major parts anyway: the parser, the runner and the reporter. The parser must be specified, it handles the specific testing language you will develop for the project. 
+
+The runner is the easiest to provide a default implementation for and the reporters that can be made available are limited only by the time you have to come up with solutions but a simple console output is easy to provide.
```
Comment (mrkita) [inline examples/README.md:33]: *for and ???

**Review by mrkita — CHANGES_REQUESTED** [review]

File: lib/rutema/core/engine.rb [hunk 15]
```diff
@@ -1,18 +1,34 @@
...
 require_relative '../version'
 
 module Rutema
-  #Rutema::Engine implements the rutema workflow.
+  ##
+  # Class implementing the inner workflow of rutema
   #
-  #It instantiates the configured parser, runner and reporter instances and wires them together via Rutema::Dispatcher
+  # This class takes care of instantiating the configured parser, runner and
+  # reporters. The runner and the reporters are connected by means of a
```
[the remark this answers is not in the document]
Comment (markuspg) [reply lib/rutema/core/engine.rb]: Replaced by your suggestion

File: lib/rutema/core/engine.rb [hunk 16]
```diff
@@ -1,18 +1,34 @@
...
+  # Dispatcher instance.
   #
-  #The full workflow is Parse->Run->Report and corresponds to one call of the Engine#run method
+  # The full workflow consists of +parse+, +run+ and +report+ and corresponds
```
[the remark this answers is not in the document]
Comment (markuspg) [reply lib/rutema/core/engine.rb]: Good suggestion, added :+1:

File: examples/README.md (hunk 13 above)
Comment (markuspg) [reply examples/README.md]: Changed :+1:

File: examples/README.md (hunk 14 above)
Comment (markuspg) [reply examples/README.md]: I reformulated the paragraph for hopefully better understandability.

File: examples/README.md (hunk 12 above)
Comment (markuspg) [reply examples/README.md]: I merged the two paragraphs into one and tried to make the distinction between the two types of configuration files more clear.

File: lib/rutema/core/engine.rb (hunk 8 above)
Comment (markuspg) [reply lib/rutema/core/engine.rb]: Applied your suggestion :+1:

File: lib/rutema/core/engine.rb (hunk 9 above)
Comment (markuspg) [reply lib/rutema/core/engine.rb]: Applied you suggestion :+1:

File: lib/rutema/parsers/xml.rb (hunk 11 above)
Comment (markuspg) [reply lib/rutema/parsers/xml.rb]: Changed :+1:

File: lib/rutema/core/engine.rb (hunk 10 above)
Comment (markuspg) [reply lib/rutema/core/engine.rb:342]: Changed :+1:

File: lib/rutema/core/engine.rb (hunk 7 above)
Comment (markuspg) [reply lib/rutema/core/engine.rb]: Changed :+1:

File: lib/rutema/core/engine.rb (hunk 6 above)
Comment (markuspg) [reply lib/rutema/core/engine.rb]: Changed :+1:

File: lib/rutema/core/engine.rb (hunk 4 above)
Comment (markuspg) [reply lib/rutema/core/engine.rb]: I went with your wording :+1:

File: lib/rutema/core/engine.rb (hunk 1 above)
Comment (markuspg) [reply lib/rutema/core/engine.rb:53]: Indeed I'm not so happy with my wording either. It's always difficult to summarize complex conditions in short sentences.

The point about the optional `test_identifier` is, that it must be specified somewhere in the configuration too. I cannot specify any `test_identifier` here, even if it were a valid one.

File: lib/rutema/core/engine.rb (hunk 5 above)
Comment (markuspg) [reply lib/rutema/core/engine.rb:228]: I'm sorry, but here the context is different. How could or should this one be improved.

The `test_identifier` name is anyhow one I'm not satisfied yet. I would rename this during the refactoring itself. This pull request is more about laying a baseline for the refactoring and getting my head around the code.

**Pull request merged**